### ISSUE TITLE
Internal XLA error when integrating multiheaded attention

### ISSUE THREAD

**Issue by jatentaki** [google/jax#7928]
# Description
I'm trying to integrate a transformer block defined in flax, using code provided [here](https://github.com/google/flax/issues/1496#issuecomment-907781279) by @levskaya. For some reason the code I'm playing with results in an internal XLA error `RuntimeError: Internal: Inconsistent reduction fusion outputs`. I assume that internal XLA errors should not surface through JAX APIs, therefore I report this here.

# Repro
Provided [here](https://colab.research.google.com/drive/191luKX_FCpakGnXkmmOg4boUv18JN0K1?usp=sharing). It appears that the bug requires that the MHA block contains both CrossAttention, MLP and LayerNorms. Removing either part usually makes it go away, although I haven't tried all possible combinations.

_Note_: I'm not sure if this is the right location to report (could go to flax, jax or xla); please reroute the issue where you find appropriate.

**Comment by levskaya** [comment]
Hey!  This does look almost certainly like an internal XLA bug - we can usually get these fixed quickly if we have an "HLO dump" of the offending computation to give the XLA devs.

Irritatingly HLO dumping got broken recently but we just released a fixed lib-tpu that you'd need to update to:
`pip install --upgrade libtpu-nightly -f https://storage.googleapis.com/jax-releases/libtpu_releases.html --no-cache`
you should get `libtpu-nightly-0.1.dev20210916` from that.

After that, if you can run your binary with the environment variable `XLA_FLAGS="--xla_dump_to=DIRECTORY_PATH"` set to a dump location you'll get a lot of HLO files named after the jax computations - send them all our way, but if you know which computation is breaking from the logs maybe that one alone will suffice.  We can test the compiler input for the bug independently.

**Comment by jatentaki** [comment]
Thanks for the answer but I'm not sure if I understand correctly. I'm not using TPUs here, why should libtpu matter? Is it just a way to get fixed HLO dumping on board (while it's still broken in libjax)? In any case I'll try that soon and let you know.

**Comment by jatentaki** [comment]
I just tried following your suggested steps but with libtpu downloaded, the notebook seems to get stuck at `import jax.numpy as jnp` showing the following trace:
![image](https://user-images.githubusercontent.com/22394841/133855951-35a0d8fc-c42a-4743-9879-4f1acc1c50a7.png)
and after ~2minutes it crashed with 
```
/usr/local/lib/python3.7/dist-packages/jax/__init__.py:27: UserWarning: cloud_tpu_init failed: ConnectionError(MaxRetryError("HTTPConnectionPool(host='metadata.google.internal', port=80): Max retries exceeded with url: /computeMetadata/v1/instance/attributes/agent-worker-number (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f18c1dec710>: Failed to establish a new connection: [Errno 110] Connection timed out'))"))
 This a JAX bug; please report an issue at https://github.com/google/jax/issues
  _warn(f"cloud_tpu_init failed: {repr(exc)}\n This a JAX bug; please report "
```
I also tried the same in a runtime with TPU available and it didn't make a difference (can it possibly need more than a minute of time to find it?).

When libtpu is not present I get 
```
2021-09-17 21:28:20.670463: F external/org_tensorflow/tensorflow/compiler/xla/parse_flags_from_env.cc:170] Could not open file ""--xla_dump_to=/xla-dumps"" to read flags for environment variable "XLA_FLAGS". (We assumed ""--xla_dump_to=/xla-dumps"" was a file name because it did not start with a "--".)
```
presumably that's what you mean by HLO dumping being broken.

**Comment by levskaya** [comment]
Sorry I just assumed you were on TPU - so ignoring all that libtpu stuff, does the dumping envvar work on a plain CPU or GPU box (whatever you're using), or do you get the same error?  Sorry for the trouble!

**Comment by jatentaki** [comment]
I was actually playing in colab when this happened, so the repro is the "realest" thing :) I just tried on CPU, which works fine. I also fought a long fight trying to get the XLA dumps, it seems to just work with `%env XLA_FLAGS=--xla_dump_to=/xla-dumps` added to the preamble and generates [those files](https://drive.google.com/file/d/1MRs4VkKBY1fij_PnjVe3USmVSYSoeK_H/view?usp=sharing). Not being familiar with this intermediate representation I can't tell if they are broken or not. I also tried older versions of jax/jaxlib (which is a mess because it disables use of flax 0.3.4 forcing me to copy-paste parts of the source into the notebook) and the operation seems to work in some older versions of jax/jaxlib (no XLA error). Is there something more I can do for the case?
